src/tdycore.c

```diff
@@ -1833,14 +1950,58 @@ PetscErrorCode TDyPreSolveSNESSolver(TDy tdy) {
   PetscFunctionReturn(0);
 }
 
+/// Resets solver when a time step is cut
+/// @param [inout] TDy struct
+///
+/// @returns 0 on success, or a non-zero error code on failure
+PetscErrorCode TDyTimeCut(TDy tdy) {
```

Comment:
Seems like this behavior should be added to our set of "polymorphic" functions.

Reply:
The way the TDycore code (outside of this PR) is implemented, we have to use `ierr = tdy->ops->time_cut(tdy);` instead of `ierr = tdy->ops->time_cut(&tdy->context, ...);` Would `ierr = tdy->ops->time_cut(tdy);` be a valid polymorphic function?

Reply:
Good point. I guess we haven't really talked about this much. When I refactored TDycore to be polymorphic, I wanted to make it clear which information was available to each function instead of just passing the `TDy`object itself, which has pretty much everything. This is an aesthetic design choice, but it can be helpful to "suggest" which data a function "should" or "shouldn't" modify.

In this case, maybe we can just pass the `TDy` object as you have done. If we later find a pattern to our approaches for cutting the timestep, we can always go back and change the function to take specific data instead of `TDy`.